tiers/checks.go

```diff
@@ -0,0 +1,157 @@
+package tiers
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/codeready-toolchain/toolchain-e2e/wait"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	providerMatchingLabels = client.MatchingLabels(map[string]string{"provider": "codeready-toolchain"})
+)
+
+func NewTierChecks(tier string) (TierChecks, error) {
+	switch tier {
+	case "basic":
+		return &basicTierChecks{}, nil
+
+	case "advanced":
+		return &advancedTierChecks{basicTierChecks: &basicTierChecks{}}, nil
+
+	case "team":
+		return &teamTierChecks{}, nil
+
+	default:
+		return nil, fmt.Errorf("no assertion implementation found for %s", tier)
+	}
+}
+
+type TierChecks interface {
+	GetInnerObjectChecks(nsType string) []innerObjectCheck
+	GetExpectedRevisions(awaitility *wait.Awaitility) Revisions
+}
+
+type basicTierChecks struct {
+}
+
+func (a *basicTierChecks) GetInnerObjectChecks(nsType string) []innerObjectCheck {
+	if nsType == "code" {
+		return []innerObjectCheck{
+			userEditRoleBinding(),
+			toolchainUserEditRole("toolchain-che-edit"),
+			toolchainUserEditRoleBinding("toolchain-che-edit"),
+			numberOfToolchainRoles(1),
+			numberOfToolchainRoleBindings(2),
+		}
+	}
+	return []innerObjectCheck{
+		userEditRoleBinding(),
+		numberOfToolchainRoles(0),
+		numberOfToolchainRoleBindings(1),
+	}
+}
+
+func (a *basicTierChecks) GetExpectedRevisions(awaitility *wait.Awaitility) Revisions {
+	revisions, err := GetRevisions(awaitility, "basic", "code", "dev", "stage")
+	require.NoError(awaitility.T, err)
+	return revisions
+}
+
+type advancedTierChecks struct {
+	*basicTierChecks
+}
+
+func (a *advancedTierChecks) GetExpectedRevisions(awaitility *wait.Awaitility) Revisions {
+	revisions, err := GetRevisions(awaitility, "advanced", "code", "dev", "stage")
+	require.NoError(awaitility.T, err)
+	return revisions
+}
+
+type teamTierChecks struct {
+}
+
+func (a *teamTierChecks) GetInnerObjectChecks(nsType string) []innerObjectCheck {
```

Comment:
do you need a pointer receiver?

Reply:
We don't have to use a pointer receiver in this case but it doesn't hurt, does it? :)
I personally prefer to use pointer receivers since it's IMO less error prone. Is there any disadvantages of using them?

Reply:
yes, it's my habit to use a pointer receiver in all methods and not use it only when there is a reason for it.